warehouse/query-core/src/main/java/datawave/query/jexl/visitors/DelayedNonEventSubTreeVisitor.java

```diff
@@ -0,0 +1,124 @@
+        this.nonEventFields = nonEventFields;
+        delayedNonEventFieldMapNodes = HashMultimap.create();
+    }
+    
+    /**
+     * Check if the node is an instance of a DelayedPredicate and if so capture all nonEvent delayed nodes, ranges, and node operators into
+     * delayedNonEventFieldMapNodes.
+     * 
+     * @param node
+     * @param data
+     * @return
+     */
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        if (ASTDelayedPredicate.instanceOf(node) || (data != null && (boolean) data)) {
+            // strip off only the delayed predicate, leave any other markers intact since they may be necessary to properly process inside the
+            // IteratorBuildingVisitor
+            JexlNode candidate = QueryPropertyMarker.getQueryPropertySource(node, ASTDelayedPredicate.class);
+            
+            if (candidate == null) {
+                // we are in a recursive situation inside an delayed predicate, evaluate the node
+                candidate = node;
+            }
+            
+            extractDelayedFields(candidate);
+            
+            // continue processing the tree in case the IteratorBuildingVisitor decided to discard leafs. This is not the most efficient or ideal
+            // but it avoids logic problems in AndIterator/OrIterator that haven't yet been fixed.
+            return super.visit(candidate, true);
+        }
+        
+        // it wasn't a delayed node, process down the tree
+        return super.visit(node, data);
+    }
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        return processDelayedFields(node, data);
+    }
+    
+    @Override
+    public Object visit(ASTEQNode node, Object data) {
+        return processDelayedFields(node, data);
+    }
+    
+    @Override
+    public Object visit(ASTNENode node, Object data) {
+        return processDelayedFields(node, data);
+    }
+    
+    private Object processDelayedFields(JexlNode node, Object data) {
+        if (data != null && (boolean) data) {
+            // continue processing the tree in case the IteratorBuildingVisitor decided to discard leafs. This is not the most efficient or ideal
+            // but because the IteratorBuildingVisitor will only identify a root if it has includes we have to look after each recursion.
+            extractDelayedFields(node);
+        }
+        
+        return super.visit(node, data);
+    }
+    
+    private void extractDelayedFields(JexlNode node) {
+        // reset the root before processing a a node
+        iteratorBuildingVisitor.resetRoot();
+        
+        // use the iterator building visitor to build a field to node map for each delayed node
+        node.jjtAccept(iteratorBuildingVisitor, null);
+        NestedIterator<Key> root = iteratorBuildingVisitor.root();
+        if (root != null) {
+            for (NestedIterator<Key> leaf : root.leaves()) {
```

Comment:
Ultimately any iterator that is created directly against the shard will be included here. Since only delayed sub trees make it to this point in the visitor, they can be checked against non-event fields since all other fields will have their values in the event and don't actually required a deferred iterator.